pkg/controller/ipam/metrics.go

```diff
@@ -0,0 +1,150 @@
+	metricSubsystemIPAM   = "ipam"
+)
+
+type AntreaIPAMMetricsHandler struct {
+	crdClient      versioned.Interface
+	ipPoolLister   crdlisters.IPPoolLister
+	ipPoolInformer v1alpha2.IPPoolInformer
+}
+
+var (
+	ipPoolUsageTotal = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      metricNamespaceAntrea,
+			Subsystem:      metricSubsystemIPAM,
+			Name:           "ippool_usage_total",
+			Help:           "Total number of available IP addresses for the address pool.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"pool_name"},
+	)
+	ipPoolUsageUsed = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      metricNamespaceAntrea,
+			Subsystem:      metricSubsystemIPAM,
+			Name:           "ippool_usage_used",
+			Help:           "Number of allocated IP addresses for the address pool.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"pool_name"},
+	)
+)
+
+func NewAntreaIPAMMetricsHandler(crdClient versioned.Interface, crdInformerFactory crdinformers.SharedInformerFactory) *AntreaIPAMMetricsHandler {
+	legacyregistry.MustRegister(ipPoolUsageTotal)
+	legacyregistry.MustRegister(ipPoolUsageUsed)
+
+	ipPoolInformer := crdInformerFactory.Crd().V1alpha2().IPPools()
+	mh := &AntreaIPAMMetricsHandler{
+		crdClient:      crdClient,
+		ipPoolInformer: ipPoolInformer,
+		ipPoolLister:   ipPoolInformer.Lister()}
+	return mh
+}
+
+func (c *AntreaIPAMMetricsHandler) Run(stopCh <-chan struct{}) {
+	c.ipPoolInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.createHandler,
+		UpdateFunc: c.updateHandler,
+		DeleteFunc: c.deleteHandler,
+	}, 0)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformer.Informer().HasSynced) {
+		return
+	}
+	<-stopCh
+}
+
+func (c *AntreaIPAMMetricsHandler) createHandler(obj interface{}) {
+	c.updateIPPoolCounters(obj)
+}
+
+func (c *AntreaIPAMMetricsHandler) updateHandler(oldObj, newObj interface{}) {
+	c.updateIPPoolCounters(newObj)
+}
+
+func (c *AntreaIPAMMetricsHandler) deleteHandler(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	if !ipPoolUsageTotal.DeleteLabelValues(ipPool.Name) {
+		klog.Warningf("Deletion of IPPool %s ippool_usage_total metric failed", ipPool.Name)
+	}
+	if !ipPoolUsageUsed.DeleteLabelValues(ipPool.Name) {
+		klog.Warningf("Deletion of IPPool %s ippool_usage_total metric failed", ipPool.Name)
+	}
+}
+
+func (c *AntreaIPAMMetricsHandler) updateIPPoolCounters(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPool.Name, c.crdClient)
+
+	if err != nil {
+		klog.Warningf("Failed to initialize allocator for IPPool %s", ipPool.Name)
+	}
+
+	// Total is fetched from allocator as here are trapped changes to CRD, e.g addition of new IPRange
+	total := allocator.Total()
+
+	// Used is gathered from IP allocation status within the CRD - as it can be set by each one of the agents
+	used := len(ipPool.Status.IPAddresses)
```

Comment:
Here I'm not sure whether we want to count reserved addresses? Those would be the addresses reserved for StatefulSet but not actually allocated for Pods.

Reply:
Makes sense that we count them. These numbers should reflect the usage, whether it's a reserved or allocated for a Pod.
Otherwise total minus used won't reflect the available IPs in the pool, right?